File: src/codes/duplicate-logic/ts/rfc_template.ts
abstract class Exporter {
  export(data: string): void {
    console.log("開始ログ");
    this.doExport(data); // フックポイント（抽象メソッド）
    console.log("完了ログ");
  }

  protected abstract doExport(data: string): void;
}

class PdfExporter extends Exporter {
  protected doExport(data: string): void {
    console.log(`PDF出力: ${data}`);
  }
}

class CsvExporter extends Exporter {
  protected doExport(data: string): void {
    console.log(`CSV出力: ${data}`);
  }
}

class XmlExporter extends Exporter {
  protected doExport(data: string): void {
    console.log(`XML出力: ${data}`);
  }
}

// 利用例
const exporter = new CsvExporter();
exporter.export("データ123");

File: src/codes/duplicate-logic/py/rfc_template.py
from abc import ABC, abstractmethod

class Exporter(ABC):
    def export(self, data: str):
        print("開始ログ")
        self.do_export(data)  # フックメソッド
        print("完了ログ")

    @abstractmethod
    def do_export(self, data: str):
        pass

class PdfExporter(Exporter):
    def do_export(self, data: str):
        print(f"PDF出力: {data}")

class CsvExporter(Exporter):
    def do_export(self, data: str):
        print(f"CSV出力: {data}")

class XmlExporter(Exporter):
    def do_export(self, data: str):
        print(f"XML出力: {data}")

# 利用例
exporter = PdfExporter()
exporter.export("データ123")
